[#139, #141, #144, #148] Milestones 페이지 컴포넌트 구조 구현

junsushin-dev opened the pull request:
## 구현의도
- 마일스톤 페이지 기본 구조 및 스타일링 적용

## 기능 흐름도, 클래스 다이어그램(선택사항)

## 사용된 기술
- styled components: props로 조건 받아서 선택적 CSS 적용

## 리뷰요청 & 논의사항 & 궁금한점
- 레이블 페이지와 마일스톤 페이지 헤더가 같아서 재사용 가능한 컴포넌트로 구현하였습니다. 

closes #139 #141 #144 #148

junsushin-dev committed:
feat: issueListItem 컴포넌트 구현

마일스톤 1개 단위 컴포넌트 구현 - 제목, 마감일, 상세, 진척도, 버튼들

junsushin-dev committed:
feat: milestonesAPI 실제 API로 연동

Mock JSON 데이터에서 실제 API 연동으로 변경

junsushin-dev committed:
feat: MilestoneList 컴포넌트 구현

개별 MilestoneListItem을 담아놓을 목적의 컴포넌트

junsushin-dev committed:
feat: MilestonesPage 구조변경 및 상태관리 추가

Header, ListNav, List 단위 구조 추가
milestones 정보 상태관리 추가

junsushin-dev committed:
feat: OpenCloseButton 컴포넌트 구현

MilestoneListNav에서 Open, Close 된 milestone의 개수 표시 용도

junsushin-dev committed:
feat: MilestoneListNav 컴포넌트 구현

Navbar 컴포넌트 선언, 스타일링, Open/Closed 개수 계산

junsushin-dev committed:
feat: NewButton 컴포넌트 구현

NewIssue, NewLabel, NewMilestone 으로 재사용하기위한 컴포넌트

junsushin-dev committed:
feat: PropertyButton 선택 시 색상변경 기능 추가

LablesPage, MilestonesPage 에서 헤더에 현재 선택된 메뉴 표시용

junsushin-dev committed:
feat: LabelMilestoneHeader 컴포넌트 추가

Labels 페이지와 Milestones 페이지 헤더가 동일하여 재사용 컴포넌트로 구현

junsushin-dev committed:
fix: milestone 모델에 isOpen 속성 추가

명세서 상 milestone은 issue와 동일하게 Open, Close 가능함
DB 모델과 API에서 isOpen 여부를 가져올 수 있도록 변경

junsushin-dev committed:
feat: MilestoneListItem 컴포넌트 스타일 적용

내용 배치, 텍스트 색상, 간격 조정

passwd10 commented on frontend/src/components/MilestoneListItem.jsx:
```diff
@@ -0,0 +1,96 @@
+}
+
+export default function MilestoneListItem({ milestone }) {
+  const { closedIssueCount, allIssueCount } = milestone;
+  const openIssueCount = allIssueCount - closedIssueCount;
+  const completionRate = Math.round(closedIssueCount / allIssueCount * 100);
+  const dueBy = new Date(milestone.dueDate).toDateString();
```
변수명이 참 이쁘네요ㅎㅎ :100:

junsushin-dev replied:
감사합니다 :)

passwd10 commented on frontend/src/pages/MilestonesPage.jsx:
```diff
@@ -1,7 +1,29 @@
 export default function MilestonesPage() {
+  const [milestones, setMilestones] = useState([]);
+
+  useEffect(async () => {
+    const newMilestones = await getMilestones();
+    setMilestones(newMilestones);
+  });
```
useEffect안에서 async await사용하면 에러나지 않나요? useEffect함수는 Promise가 아닌 함수만 반환해야한다고 하네요! async await 함수를 따로 만들어서 useEffect안에서 호출하는것이 best practice 인 것 같습니다 :smile: 

- https://im-developer.tistory.com/155
- https://athilog.github.io/react-native/useEffect%20aync%EC%B2%98%EB%A6%AC/

 글 참고하시면 좋을 것 같습니다~

junsushin-dev replied:
useEffect 함수 자체가 await로 대기하고 있을 것 같아서 인서님 말씀이 맞는 것 같네요! 다시 분리하고 오겠습니다. 좋은 코멘트 감사드려요. 😀

junsushin-dev replied:
고쳤습니다! 지금 보니 async 인것과 또 useEffect에 두번째 인자를 안넣어줘서 계속 렌더링이 되고 useEffect로 비동기 호출을 반복하고 있더라구요... 인서님이 찾아주셔서 정말 다행입니다!

passwd10 reviewed (commented):
재활용을 고려해 컴포넌트를 잘 분리하신 것 같습니다! :+1: 

review 확인부탁드립니다~

junsushin-dev committed:
fix: useEffect에서 비동기함수를 내부 선언으로 변경

useEffect 함수 자체가 비동기함수이면 오류가 발생 가능
useEffect 두번째 인자로 빈 array 추가하여 반복 렌더링하지 않도록 수정

junsushin-dev committed:
Merge branch 'develop' into feature/milestone-list

dhyun2 reviewed (approved):
깔끔해서 보기 편한것 같습니다! 수고 많으셨습니당 😊
